graphql/src/main/java/com/scalar/db/graphql/datafetcher/TransactionInstrumentation.java

```diff
@@ -0,0 +1,145 @@
+    boolean isCommitTrue =
+        commitArg != null
+            && commitArg.getValue() != null
+            && ((BooleanValue) commitArg.getValue()).isValue();
+    if (txIdArg == null) {
+      // start a new transaction
+      try {
+        DistributedTransaction transaction = transactionManager.start();
+        transactionMap.put(transaction.getId(), transaction);
+        graphQLContext.put(Constants.CONTEXT_TRANSACTION_KEY, transaction);
+        if (isCommitTrue) {
+          return new TransactionCommitInstrumentationContext(transaction);
+        }
+      } catch (TransactionException e) {
+        throw new AbortExecutionException("Failed to start transaction.", e);
+      }
+    } else {
+      // continue an existing transaction
+      String txId = ((StringValue) txIdArg.getValue()).getValue();
+      if (txId != null) {
+        DistributedTransaction transaction = transactionMap.get(txId);
+        if (transaction == null) {
+          throw new AbortExecutionException(
+              "The specified "
+                  + Constants.TRANSACTION_DIRECTIVE_TX_ID_ARGUMENT_NAME
+                  + " "
+                  + txId
+                  + " does not exist.");
+        }
+        graphQLContext.put(Constants.CONTEXT_TRANSACTION_KEY, transaction);
+        if (isCommitTrue) {
+          return new TransactionCommitInstrumentationContext(transaction);
+        }
+      }
+    }
+    return SimpleInstrumentationContext.noOp();
+  }
+
+  @Override
+  public CompletableFuture<ExecutionResult> instrumentExecutionResult(
+      ExecutionResult executionResult, InstrumentationExecutionParameters parameters) {
+    GraphQLContext graphQLContext = parameters.getGraphQLContext();
+    DistributedTransaction transaction = graphQLContext.get(Constants.CONTEXT_TRANSACTION_KEY);
+    if (transaction == null) {
+      return CompletableFuture.completedFuture(executionResult);
+    }
+
+    Map<Object, Object> currentExt = executionResult.getExtensions();
+    Map<Object, Object> withTxExt =
+        new LinkedHashMap<>(currentExt == null ? Collections.emptyMap() : currentExt);
+    withTxExt.put(
+        EXTENSIONS_TRANSACTION_KEY, ImmutableMap.of(EXTENSIONS_TX_ID_KEY, transaction.getId()));
+
+    return CompletableFuture.completedFuture(
+        new ExecutionResultImpl(executionResult.getData(), executionResult.getErrors(), withTxExt));
+  }
+
+  @VisibleForTesting
+  static class TransactionCommitInstrumentationContext
+      implements InstrumentationContext<ExecutionResult> {
+
+    private final DistributedTransaction transaction;
+
+    public TransactionCommitInstrumentationContext(DistributedTransaction transaction) {
+      this.transaction = transaction;
+    }
+
+    @Override
+    public void onDispatched(CompletableFuture<ExecutionResult> result) {}
+
+    @Override
+    public void onCompleted(ExecutionResult result, Throwable t) {
+      if (t != null) {
```

Comment:
It looks like the transaction is not committed if `t` is not null, even though specifying `commit=true` in the transaction directive. In this case, we need to call `abort` explicitly, right?

Reply:
Fixed in the latest commit. Thank you!